src/rpc/connection.rs

```diff
@@ -627,13 +623,15 @@ impl CircularQueue {
         &self,
         id: u32,
         status: libc::c_int,
+        flags: u32,
         meta_data_length: usize,
         data_lenght: usize,
     ) -> Result<(), Box<dyn std::error::Error>> {
         unsafe {
             let callback = self.callbacks[id as usize];
             (*(callback as *mut OperationCallback)).state = CallbackState::Done;
```

Comment:
why cause inconsistency of state? But I think it's a duplicate

Reply:
In concurrent case, while clean_up set state as Empty from Done, the state may be set as Done at wait_for_callbakc() again.

Reply:
> set state as Done twice at both response() and wait_for_callback(). It may cause inconsistency of state?

I removed duplicated state change and put it behind the code reading data from callback, so that make sure the callback will not be written by other calls.

Reply:
> > set state as Done twice at both response() and wait_for_callback(). It may cause inconsistency of state?
> 
> I removed duplicated state change and put it behind the code reading data from callback, so that make sure the callback will not be written by other calls.

but clean_up() happen in response(), it is bette to set the state before clean_up.